tests/Unit/CrudPanel/CrudPanelFieldsTest.php

```diff
@@ -0,0 +1,539 @@
+        $this->assertEmpty($this->crudPanel->create_fields);
+    }
+
+    public function testBeforeField()
+    {
+        $this->markTestIncomplete('Not correctly implemented');
+
+        $this->crudPanel->addFields($this->threeTextFieldsArray);
+
+        // TODO: fix the before field method not preserving field keys
+        $this->crudPanel->beforeField('field1');
+
+        $createKeys = array_keys($this->crudPanel->create_fields);
+        $this->assertEquals($this->expectedThreeTextFieldsArray['field3'], $this->crudPanel->create_fields[$createKeys[0]]);
+        $this->assertEquals(['field3', 'field1', 'field2'], $createKeys);
+
+        $updateKeys = array_keys($this->crudPanel->update_fields);
+        $this->assertEquals($this->expectedThreeTextFieldsArray['field3'], $this->crudPanel->update_fields[$updateKeys[0]]);
+        $this->assertEquals(['field3', 'field1', 'field2'], $updateKeys);
+    }
+
+    public function testBeforeFieldCreateForm()
+    {
+        $this->markTestIncomplete('Not correctly implemented');
+
+        $this->crudPanel->addFields($this->threeTextFieldsArray, 'create');
+
+        // TODO: fix the before field method not preserving field keys
+        $this->crudPanel->beforeField('field1');
+
+        $createKeys = array_keys($this->crudPanel->create_fields);
+        $this->assertEquals($this->expectedThreeTextFieldsArray['field3'], $this->crudPanel->create_fields[$createKeys[0]]);
+        $this->assertEquals(['field3', 'field1', 'field2'], $createKeys);
+
+        $this->assertEmpty($this->crudPanel->update_fields);
+    }
+
+    public function testBeforeFieldUpdateForm()
+    {
+        $this->markTestIncomplete('Not correctly implemented');
+
+        $this->crudPanel->addFields($this->threeTextFieldsArray, 'update');
+
+        // TODO: fix the before field method not preserving field keys
+        $this->crudPanel->beforeField('field1');
+
+        $updateKeys = array_keys($this->crudPanel->update_fields);
+        $this->assertEquals($this->expectedThreeTextFieldsArray['field3'], $this->crudPanel->update_fields[$updateKeys[0]]);
+        $this->assertEquals(['field3', 'field1', 'field2'], $updateKeys);
+
+        $this->assertEmpty($this->crudPanel->create_fields);
+    }
+
+    public function testBeforeFieldForDifferentFieldsInCreateAndUpdate()
+    {
+        $this->markTestIncomplete('Not correctly implemented');
+
+        $this->crudPanel->addFields($this->threeTextFieldsArray, 'create');
+        $this->crudPanel->addFields($this->twoTextFieldsArray, 'update');
+
+        // TODO: fix the before field method not preserving field keys
+        // TODO: in the case of the create form, this will move the 'field3' field before the 'field1' field, but for
+        //       the update form, it will move the 'field2' field before the 'field1' field. should it work like this?
+        $this->crudPanel->beforeField('field1');
```

Comment:
@tabacitu, any thoughts on this?

Reply:
Completely agree :-) A second parameter for ```beforeField()``` makes a lot of sense, if that's your suggestion :-)